apps/src/p5lab/spritelab/commands/criterionCommands.js

```diff
@@ -27,25 +27,53 @@ export const commands = {
   // Return true if any sprite began speaking.
   anySpriteSpeaks() {
     const spriteIds = this.getSpriteIdsInUse();
-    let result = false;
-    for (let i = 0; i < spriteIds.length; i++) {
-      if (commands.spriteSpeechRenderedThisFrame.call(this, i)) {
-        result = true;
-      }
-    }
-    return result;
+    return (
+      spriteIds.filter(_spriteId =>
+        commands.spriteSpeechRenderedThisFrame.call(this, _spriteId)
+      ).length > 0
+    );
+  },
+
+  // Return true if the specified sprite began speaking
+  // and the text is not an empty string.
+  strictSpriteSpeechRenderedThisFrame(spriteId) {
+    return (
+      this.getLastSpeechBubbleForSpriteId(spriteId)?.renderFrame ===
+        this.currentFrame() &&
+      this.getLastSpeechBubbleForSpriteId(spriteId)?.text
+    );
+  },
+
+  // Return true if any sprite began speaking
+  // and the text is not an empty string.
+  strictAnySpriteSpeaks() {
+    const spriteIds = this.getSpriteIdsInUse();
+    return (
+      spriteIds.filter(_spriteId =>
```

Comment:
ah, sorry, i used the underscore convention (`_spriteId`) because i thought it would be an unused variable -- in the previous implementation, i thought we were using index, not ID, but i must've misread it. this variable should be `spriteId` or just `id` (when you're iterating over a list of IDs, `id` is usually sufficient context)

Reply:
You didn't misread it; there was a bug (see other comment)! I was wondering about this underscore convention. I'll make one last update to rename to `id`, then assume you're still good to approve.